src/actions.ml

```diff
@@ -2075,3 +2086,87 @@ let push_action ~bot_info ~base_ref ~commits_msg =
     else Lwt.return ()
   in
   Lwt_list.iter_s commit_action commits_msg
+
+let days_elapsed ts =
+  (* Yes, I know this is wrong because of DST and black holes but it should
+     still be correct enough *)
+  Float.to_int ((Unix.time () -. ts) /. (3600. *. 24.))
+
+let rec apply_throttle len action args =
+  if List.is_empty args || len <= 0 then Lwt.return ()
+  else
+    let args, rem = List.split_n args len in
+    Lwt_list.map_p action args >>= fun ans ->
+    let n = List.count ~f:(fun b -> b) ans in
+    apply_throttle (len - n) action rem
+
+let apply_after_label ~bot_info ~owner ~repo ~after ~label ~action ?throttle () =
+  GitHub_queries.get_open_pull_requests_with_label ~bot_info ~owner ~repo ~label
+  >>= function
+  | Ok prs ->
+    let iter (pr_id, pr_number) =
+      GitHub_queries.get_pull_request_label_timeline ~bot_info ~owner ~repo ~pr_number >>= function
+      | Ok timeline ->
+        let find (set, name, ts) =
+          if set && String.equal name label then Some ts else None
+        in
+        (* Look for most recent label setting *)
+        let timeline = List.rev timeline in
+        let days = match List.find_map ~f:find timeline with
+        | None ->
+          (* even with a race condition it cannot happen *)
+          failwith (f "Anomaly: Label \"%s\" absent from timeline of PR #%i" label pr_number)
+        | Some ts -> days_elapsed ts
+        in
+        if days >= after then
+          action pr_id pr_number
+        else Lwt.return false
+      | Error e ->
+        Lwt_io.print (f "Error: %s\n" e) >>= fun () -> Lwt.return false
+    in
+    begin match throttle with
+    | None -> Lwt_list.iter_p (fun v -> iter v >>= fun _ -> Lwt.return ()) prs
+    | Some throttle ->
+      apply_throttle throttle iter prs
+    end
+  | Error err ->
+    Lwt_io.print (f "Error: %s\n" err)
+
+let coq_check_needs_rebase_pr ~bot_info ~owner ~repo ~warn_after ~close_after ~throttle =
+  let label = "needs: rebase" in
+  GitHub_queries.get_label ~bot_info ~owner ~repo ~label:"stale" >>= function
+  | Ok stale_id ->
+    let action pr_id pr_number =
+      GitHub_queries.get_pull_request_labels ~bot_info ~owner ~repo ~pr_number >>= function
```

Comment:
Isn't this query a bit redundant with the one that gets the label timeline in `apply_after_labels`? Wouldn't it make sense for the latter to call the action with the timeline as an extra argument?